File: aoe2x/js_simulation/battle_setup.py
"""Recorder purchase-cost convention shared by runtime profiles and comparisons."""

RESOURCES = ("food", "wood", "gold")


def effective_cost(row):
    """Fully discounted cost per physical unit (recorder policy at 2cced342).

    Saxon maximum conditional discount: four controlled TCs/Castles, 20%.
    Blackwood Archer purchase trains two units; divide after resource rounding.
    """
    purchase = {r: row[f"final_cost_{r}"] for r in RESOURCES}
    multiplier = 0.8 if row["civ_name"] == "Saxons" and row["unit_class"] in (0, 6, 44) else 1
    batch = 2 if row["unit_master"] in (2579, 2581) else 1
    costs = {r: round(value * multiplier) / batch for r, value in purchase.items()}
    return costs, dict(referencePurchaseCost=purchase, unitsPerPurchase=batch,
                      conditionalMultiplier=multiplier,
                      conditionalContext="four controlled Town Centers/Castles; maximum 20% discount"
                      if multiplier != 1 else None)

File: aoe2x/js_simulation/src/battle-setup.js
import { isRangedClass } from "./lab-scenario.js";

// BALANCE_POLICY.md, September 22, 2026. effectiveCost is the fully
// discounted purchase cost of one physical unit, including batch division.
export function deriveRecordedBattleSetup(side2, side3, { cap = 27 } = {}) {
  const weightedCost = ({ effectiveCost: { food, wood, gold } }) => (
    food + 0.9 * wood + 1.1 * gold
  );
  const cost2 = weightedCost(side2);
  const cost3 = weightedCost(side3);
  const n2 = cost2 <= cost3 ? cap : Math.max(1, Math.round(cap * Math.sqrt(cost3 / cost2)));
  const n3 = cost3 <= cost2 ? cap : Math.max(1, Math.round(cap * Math.sqrt(cost2 / cost3)));
  const ranged2 = isRangedClass(side2.class);
  const ranged3 = isRangedClass(side3.class);
  const rangedOwner = ranged2 === ranged3 ? null : (ranged2 ? 2 : 3);
  const rangedArmyCost = rangedOwner === 2 ? n2 * cost2 : n3 * cost3;
  const player4Count = rangedOwner === null ? 0 : Math.round(
    Math.max(5, Math.min(10, 5 + (rangedArmyCost - 1000) / 1800)),
  );
  return { n2, n3, cost2, cost3, player4Count, rangedOwner,
    policy: "geometric_full_discount_weighted_resources_v3",
    bufferPolicy: "fielded_weighted_cost_v1" };
}
